File: templates/docker/python/otel_setup.py
import os
import logging
import sys
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

def init_telemetry(service_name: str):
    provider = TracerProvider()
    
    if os.environ.get('K_SERVICE'):
        try:
            from opentelemetry.exporter.cloud_trace import CloudTraceSpanExporter
            import google.cloud.logging
            exporter = CloudTraceSpanExporter()
            client = google.cloud.logging.Client()
            client.setup_logging()
        except ImportError:
            logging.basicConfig(level=logging.INFO)
            logging.warning("GCP OTel/Logging packages not found. Falling back to local logging.")
            exporter = None
    else:
        try:
            from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
            endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4317")
            exporter = OTLPSpanExporter(endpoint=endpoint)
            logging.basicConfig(level=logging.INFO, stream=sys.stdout)
        except ImportError:
            logging.basicConfig(level=logging.INFO, stream=sys.stdout)
            logging.warning("OTLP Exporter not found. Tracing disabled.")
            exporter = None

    if exporter:
        provider.add_span_processor(BatchSpanProcessor(exporter))
        trace.set_tracer_provider(provider)
    
    # Generic error handler uncaught exceptions
    def handle_exception(exc_type, exc_value, exc_traceback):
        if issubclass(exc_type, KeyboardInterrupt):
            sys.__excepthook__(exc_type, exc_value, exc_traceback)
            return
        logger = logging.getLogger(service_name)
        logger.error("Uncaught exception", exc_info=(exc_type, exc_value, exc_traceback))

    sys.excepthook = handle_exception
    
    return logging.getLogger(service_name)

File: templates/docker/node/otelSetup.js
const { NodeSDK } = require('@opentelemetry/sdk-node');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');

function initTelemetry(serviceName) {
  let traceExporter = null;
  let logger;

  if (process.env.K_SERVICE) {
    // Cloud Run Environment
    try {
      const { TraceExporter } = require('@google-cloud/opentelemetry-cloud-trace-exporter');
      traceExporter = new TraceExporter();
      
      const { Logging } = require('@google-cloud/logging');
      const logging = new Logging();
      const log = logging.log(serviceName);
      
      logger = {
        info: (msg) => { console.log(msg); log.write(log.entry(msg)); },
        error: (msg, err) => { console.error(msg, err); log.write(log.entry({ severity: 'ERROR' }, msg)); }
      };
    } catch (e) {
      console.warn("GCP OTel/Logging packages not found. Falling back to local logging.");
      logger = {
        info: (msg) => console.log(`[INFO] ${msg}`),
        error: (msg, err) => console.error(`[ERROR] ${msg}`, err)
      };
    }
  } else {
    // Local Environment (Jaeger)
    try {
      const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
      traceExporter = new OTLPTraceExporter({
        url: process.env.OTEL_EXPORTER_OTLP_ENDPOINT || 'grpc://localhost:4317',
      });
      logger = {
        info: (msg) => console.log(`[INFO] ${msg}`),
        error: (msg, err) => console.error(`[ERROR] ${msg}`, err)
      };
    } catch (e) {
      console.warn("OTLP Exporter not found. Tracing disabled.");
      logger = {
          info: (msg) => console.log(`[INFO] ${msg}`),
          error: (msg, err) => console.error(`[ERROR] ${msg}`, err)
      };
    }
  }

  if (traceExporter) {
    const sdk = new NodeSDK({
      traceExporter,
      instrumentations: [getNodeAutoInstrumentations()]
    });

    try {
        sdk.start();
    } catch (err) {
        console.error("Error starting OTEL NodeSDK", err);
    }

    // Graceful shutdown
    process.on('SIGTERM', () => {
      sdk.shutdown()
        .then(() => logger.info('Tracing terminated'))
        .catch((error) => logger.error('Error terminating tracing', error))
        .finally(() => process.exit(0));
    });
  }
  
  // Generic error handler
  process.on('uncaughtException', (error) => {
    logger.error('Uncaught Exception', error);
  });
  
  process.on('unhandledRejection', (reason, promise) => {
    logger.error(`Unhandled Rejection at: ${promise}, reason: ${reason}`);
  });

  return logger;
}

module.exports = { initTelemetry };
